controller/share_manager_controller.go

```diff
@@ -0,0 +1,626 @@
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	if c.queue.NumRequeues(key) < maxRetries {
+		c.logger.WithError(err).Warnf("Error syncing Longhorn share manager %v", key)
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	c.logger.WithError(err).Warnf("Dropping Longhorn share manager %v out of the queue", key)
+	c.queue.Forget(key)
+	utilruntime.HandleError(err)
+}
+
+func (c *ShareManagerController) cleanupShareManagerPod(sm *longhorn.ShareManager) error {
+	if err := c.ds.DeletePod(sm.Name); err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	return nil
+}
+
+func (c *ShareManagerController) syncShareManager(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "fail to sync %v", key)
+	}()
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if namespace != c.namespace {
+		return nil
+	}
+
+	sm, err := c.ds.GetShareManager(name)
+	if err != nil {
+		if !datastore.ErrorIsNotFound(err) {
+			c.logger.WithField("shareManager", name).WithError(err).Error("Failed to retrieve share manager from datastore")
+			return err
+		}
+
+		c.logger.WithField("shareManager", name).Debug("Can't find share manager, may have been deleted")
+		return nil
+	}
+	log := getLoggerForShareManager(c.logger, sm)
+
+	// check if we need to claim ownership
+	if sm.Status.OwnerID != c.controllerID {
+		if !c.isResponsibleFor(sm) {
+			return nil
+		}
+
+		// on owner ship transfer we set the state to unknown,
+		// which leads to cleaning up any of the old pods
+		sm.Status.OwnerID = c.controllerID
+		sm.Status.State = types.ShareManagerStateUnknown
+		sm, err = c.ds.UpdateShareManagerStatus(sm)
+		if err != nil {
+			if apierrors.IsConflict(errors.Cause(err)) {
+				return nil
+			}
+			return err
+		}
+		log.Infof("Share manager got new owner %v", c.controllerID)
+		log = getLoggerForShareManager(c.logger, sm)
+	}
+
+	if sm.DeletionTimestamp != nil {
+		if err := c.cleanupShareManagerPod(sm); err != nil {
+			return err
+		}
+		return c.ds.RemoveFinalizerForShareManager(sm)
+	}
+
+	// update at the end, after the whole reconcile loop
+	existingShareManager := sm.DeepCopy()
+	defer func() {
+		if !reflect.DeepEqual(existingShareManager.Status, sm.Status) {
+			// we end up overwriting any prior errors
+			// but if the update works, the share manager will be
+			// enqueued again anyway
+			_, err = c.ds.UpdateShareManagerStatus(sm)
+		}
+
+		if apierrors.IsConflict(errors.Cause(err)) {
+			log.WithError(err).Debug("Requeue share manager due to conflict")
+			c.enqueueShareManager(sm)
+			err = nil
+		}
+	}()
+
+	if err = c.syncShareManagerPod(sm); err != nil {
+		return err
+	}
+
+	if sm.Status.State == types.ShareManagerStateRunning {
+		// running is once the pod is in ready state
+		// which means the nfs server is up and running with the volume attached
+		// the cluster service ip doesn't change for the lifetime of the volume
+		if sm.Status.Endpoint == "" {
```

Comment:
It's better to use level based assignment rather than edge triggered assignment for Endpoint, which is similar to https://github.com/longhorn/longhorn-manager/pull/734